src/WebJobs.Script/Description/FunctionInvokerBase.cs

```diff
@@ -81,7 +85,99 @@ protected bool InitializeFileWatcherIfEnabled()
+            startedEvent = new FunctionStartedEvent(functionExecutionContext.InvocationId, Metadata);
+            _metrics.BeginEvent(startedEvent);
+
+            try
+            {
+                TraceWriter.Info($"Function started (Id={invocationId})");
+
+                invokeTask = InvokeInternal(parameters);
+                invokeTask.Wait(timeoutOrShutdownToken);
+
+                TraceWriter.Info($"Function completed (Success, Id={invocationId})");
+            }
+            catch (OperationCanceledException ex) when (ex.CancellationToken == timeoutOrShutdownToken)
+            {
+                string reason = "Host is Stopping";
+
+                // TODO: What's the best way to do this?
+                var stateProp = typeof(JobHost).GetField("_state", System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.Instance);
+                int hostState = (int)stateProp.GetValue(Host);
+
+                // 3 indicates StoppingOrStopped. If that's the case, just let it go.
+                // Otherwise it's a timeout.
+                if (hostState != 3)
+                {
+                    reason = "Timeout";
+                    ex.Data.Add("FunctionTimeout", true);
+                }
+
+                LogFunctionFailed(startedEvent, $"Failure: {reason}", invocationId);
+                throw;
+            }
+            catch (AggregateException ex)
+            {
+                ExceptionDispatchInfo exInfo = null;
+
+                // If there's only a single exception, rethrow it by itself
+                Exception singleEx = ex.Flatten().InnerExceptions.SingleOrDefault();
+                if (singleEx != null)
+                {
+                    exInfo = ExceptionDispatchInfo.Capture(singleEx);
+                }
+                else
+                {
+                    exInfo = ExceptionDispatchInfo.Capture(ex);
+                }
+
+                LogFunctionFailed(startedEvent, "Failure", invocationId);
+                exInfo.Throw();
+            }
+            catch
+            {
+                LogFunctionFailed(startedEvent, "Failure", invocationId);
+                throw;
+            }
+            finally
+            {
+                if (startedEvent != null)
+                {
+                    _metrics.EndEvent(startedEvent);
+                }
+            }
+
+            return Task.FromResult(0);
+        }
+
+        private void LogFunctionFailed(FunctionStartedEvent startedEvent, string resultString, string invocationId)
+        {
+            if (startedEvent != null)
+            {
+                startedEvent.Success = false;
+            }
+
+            string message = $"Function completed ({resultString}";
+            if (string.IsNullOrEmpty(invocationId))
+            {
+                TraceWriter.Error($"{message})");
+            }
+            else
+            {
+                TraceWriter.Error($"{message}, Id={invocationId})");
+            }
```

Comment:
nit:
Can we simplify the above by having something like:
`TraceWriter.Error($"Function completed ({"blah"}, Id={inid ?? Guid.Empty.ToString()})")`
or
`TraceWriter.Error($"Function completed ({"blah"}, Id={inid ?? "0"})")`